### PR TITLE
Add support for HPU profiler

### PR DESCRIPTION
**Pull request by jerome-habana**
## What does this PR do?
This PR adds support for HPU profiler to trace the hpu device and host events

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source-pytorch/tuning/profiler_advanced.rst [hunk 1]
```diff
@@ -72,3 +72,121 @@ Make sure the code is running while you are trying to capture the traces. It wil
 4: View the profiling logs
 ==========================
 Once the capture is finished, the page will refresh and you can browse through the insights using the **Tools** dropdown at the top left
+
+
+
+################################################
+Find bottlenecks in your code on HPU (advanced)
```
Comment (awaelchli) [inline docs/source-pytorch/tuning/profiler_advanced.rst:79]: @Felonious-Spellfire For review of docs structure here :)

File: docs/source-pytorch/accelerators/hpu_basic.rst [hunk 2]
```diff
@@ -114,3 +114,4 @@ Known limitations
 
 * `Habana dataloader <https://docs.habana.ai/en/latest/PyTorch_User_Guide/PyTorch_User_Guide.html#habana-data-loader>`__ is not supported.
 * :func:`torch.inference_mode` is not supported
+* :func:`HPUProfiler.summary()` is not supported
```
Comment (kaushikb11) [inline docs/source-pytorch/accelerators/hpu_basic.rst:117]: We don't support it for XLA as well. Not exactly a limitation for the user

File: docs/source-pytorch/tuning/profiler_advanced.rst [hunk 3]
```diff
@@ -72,3 +72,121 @@ Make sure the code is running while you are trying to capture the traces. It wil
...
+    from pytorch_lightning.profilers import HPUProfiler
+
+    profiler = HPUProfiler(filename="perf-logs")
+    trainer = Trainer(profiler=profiler)
+
+.. note::
+    Since HPUProfiler extends PyTorch Profiler, when using the HPUProfiler, wall clock time will not not be representative of the true wall clock time.
+    This is due to forcing profiled operations to be measured synchronously, when many HPU ops happen asynchronously.
+    It is recommended to use this Profiler to find bottlenecks/breakdowns, however for end to end wall clock time use
+    the ``SimpleProfiler``.
```
Comment (kaushikb11) [inline docs/source-pytorch/tuning/profiler_advanced.rst:156]: The note is repeated.

File: src/pytorch_lightning/profilers/hpu.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        group_by_input_shapes: bool = False,
+        export_to_chrome: bool = True,
+        row_limit: int = 20,
+        sort_by_key: Optional[str] = None,
+        record_module_names: bool = True,
+        **profiler_kwargs: Any,
+    ) -> None:
+        """This profiler uses PyTorch Profiler and lets you inspect the cost of.
+
+        different operators inside your model - both on the CPU and HPU
```
Comment (kaushikb11) [inline src/pytorch_lightning/profilers/hpu.py:55]:
```suggestion
        """This profiler subclasses the PyTorch Profiler and lets you inspect the cost of
        different operators inside your model - both on the CPU and HPU devices.
```

File: src/pytorch_lightning/profilers/hpu.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+        if self.profiler is not None and any(action_name.endswith(func) for func in self.STEP_FUNCTIONS):
+            if self._schedule is not None:
+                self._schedule.pre_step(action_name)
+
+            # the default schedule requires a minimum of 5 steps to properly work: `wait=1, warmup=1, active=3`.
+            # otherwise, this will raise a `segmentation fault`.
+            if self._should_override_schedule():
+                warning_cache.warn(
+                    "The PyTorch Profiler default schedule will be overridden as there is not enough "
```
Comment (kaushikb11) [inline src/pytorch_lightning/profilers/hpu.py:141]:
```suggestion
                    "The HPUProfiler default schedule will be overridden as there is not enough "
```

File: docs/source-pytorch/tuning/profiler_advanced.rst (hunk 3 above)
Comment (jerome-habana) [reply docs/source-pytorch/tuning/profiler_advanced.rst]: updated

File: examples/pl_hpu/hpu_profiler_example.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    def configure_optimizers(self):
+        return torch.optim.Adam(self.parameters(), lr=1e-3)
+
+
+class SimpleMNISTDataModule(LightningDataModule):
+    def setup(self, stage):
+        transform = transforms.Compose([transforms.ToTensor()])
+        self.mnist_train = MNIST(os.getcwd(), train=True, download=True, transform=transform)
+        self.mnist_test = MNIST(os.getcwd(), train=False, download=True, transform=transform)
```
Comment (akihironitta) [inline examples/pl_hpu/hpu_profiler_example.py:65]: `self.mnist_test` seems unused.
```suggestion
```

File: examples/pl_hpu/hpu_profiler_example.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        self.layer_1 = torch.nn.Linear(28 * 28, 10)
+
+    def forward(self, x):
+        return torch.relu(self.layer_1(x.view(x.size(0), -1)))
+
+    def cross_entropy_loss(self, logits, labels):
+        return F.nll_loss(logits, labels)
+
+    def training_step(self, train_batch, batch_idx):
+        x, y = train_batch
```
Comment (akihironitta) [inline examples/pl_hpu/hpu_profiler_example.py:52]: nit
```suggestion
    def training_step(self, batch, batch_idx):
        x, y = batch
```

File: src/pytorch_lightning/CHANGELOG.md [hunk 8]
```diff
@@ -48,6 +48,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added an option to configure the signal SLURM sends when a job is preempted or requeued ([#14610](https://github.com/Lightning-AI/lightning/issues/14610))
 
 
+- Added support for `HPUProfiler` ([#14773](https://github.com/Lightning-AI/lightning/issues/14773))
```
Comment (akihironitta) [inline src/pytorch_lightning/CHANGELOG.md:51]: nit
```suggestion
- Added support for `HPUProfiler` ([#14773](https://github.com/Lightning-AI/lightning/pull/14773))
```

File: docs/source-pytorch/tuning/profiler_advanced.rst [hunk 9]
```diff
@@ -72,3 +72,115 @@ Make sure the code is running while you are trying to capture the traces. It wil
...
+*****************************
+Visualize profiled operations
+*****************************
+To visualize the profiled operations, enable **export_to_chrome** in the :class:`~pytorch_lightning.profilers.hpu.HPUProfiler` (Default: True).
+
+.. code-block:: python
+
+    from pytorch_lightning.profilers import HPUProfiler
+
+    profiler = HPUProfiler()
```
Comment (otaj) [inline docs/source-pytorch/tuning/profiler_advanced.rst:163]: I know it's a default value for the moment, but then the code block doesn't really show anything.

```suggestion
    profiler = HPUProfiler(export_to_chrome=True)
```

File: examples/pl_hpu/hpu_profiler_example.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    def train_dataloader(self):
+        return DataLoader(self.mnist_train, batch_size=16, num_workers=1)
+
+
+if __name__ == "__main__":
+    data_module = SimpleMNISTDataModule()
+    model = SimpleMNISTModel()
+
+    trainer = Trainer(
+        profiler=HPUProfiler(),
```
Comment (otaj) [inline examples/pl_hpu/hpu_profiler_example.py:76]: I know it's an example, but this will fail, if `_KINETO_AVAILABLE == False`, since we don't have a name `HPUProfiler`. I'd say it is either:
* perfectly fine to not have that `_KINETO_AVAILABLE` check at the top since it's an example of HPU functionality and as such it is understood that HPU is available.
* or have the code full of such checks so that the example itself works even on non-HPU HW, but for example prints out a warning

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/profilers/hpu.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        self._schedule: Optional[ScheduleWrapper] = None
+        self._profiler_kwargs = profiler_kwargs
+
+        if _KINETO_AVAILABLE:
+            self._init_kineto(profiler_kwargs)
+
+        if self._sort_by_key not in self.AVAILABLE_SORT_KEYS:
+            raise MisconfigurationException(
+                f"Found sort_by_key: {self._sort_by_key}. Should be within {self.AVAILABLE_SORT_KEYS}. "
+            )
```
Comment (akihironitta) [inline src/pytorch_lightning/profilers/hpu.py:114]: Could we remove these lines because they are the exact same as in `PyTorchProfiler.__init__()` https://github.com/Lightning-AI/lightning/blob/9fc4ff3278ac360f2671549415f6b11689a263ed/src/pytorch_lightning/profilers/pytorch.py#L300-L317?

File: src/pytorch_lightning/trainer/setup.py [hunk 12]
```diff
@@ -135,6 +136,7 @@ def init_profiler(trainer: Any, profiler: Optional[Union[Profiler, str]]) -> Non
             "advanced": AdvancedProfiler,
             "pytorch": PyTorchProfiler,
             "xla": XLAProfiler,
+            "hpu": HPUProfiler,
```
Comment (akihironitta) [inline src/pytorch_lightning/trainer/setup.py:139]: Since `HPUProfiler` subclasses `PyTorchProfiler` (unlike `XLAProfiler` subclasses `Profiler`), would it make sense to have `Trainer(profiler="pytorch", accelerator="hpu")` pick `HPUProfiler` under the hood so that the code stays accelerator-agnostic? (in a seperate PR)

Comment (jerome-habana) [reply src/pytorch_lightning/trainer/setup.py:139]: Nice suggestion. This would become applicable for all strategies that are inherited too. If we are able to upstream our changes to PyTorch, then there can be seamless switch in the trainer class as you have pointed.

File: src/pytorch_lightning/profilers/hpu.py (hunk 11 above)
Comment (kaushikb11) [reply src/pytorch_lightning/profilers/hpu.py:114]: @jerome-habana

File: examples/pl_hpu/hpu_profiler_example.py (hunk 10 above)
Comment (kaushikb11) [reply examples/pl_hpu/hpu_profiler_example.py:75]: @jerome-habana

**Review by kaushikb11 — APPROVED** [review]

File: examples/pl_hpu/hpu_profiler_example.py (hunk 10 above)
Comment (jerome-habana) [reply examples/pl_hpu/hpu_profiler_example.py]: Modified to run with cpu if kineto is unavailable

File: src/pytorch_lightning/profilers/hpu.py (hunk 11 above)
Comment (jerome-habana) [reply src/pytorch_lightning/profilers/hpu.py]: updated

**Comment by kaushikb11** [comment]
@Borda @carmocca any idea why the Circle CI github workflow is failing?

**Comment by carmocca** [comment]
It does not work for forks

File: docs/source-pytorch/tuning/profiler_advanced.rst [hunk 13]
```diff
@@ -72,3 +72,115 @@ Make sure the code is running while you are trying to capture the traces. It wil
...
+################################################
+Find bottlenecks in your code on HPU (advanced)
+################################################
+**Audience**: Users who want to profile their HPU models to find bottlenecks and improve performance.
+
+----
+
+**************************
+Profile HPU models
+**************************
```
Comment (Borda) [inline docs/source-pytorch/tuning/profiler_advanced.rst:87]:
```suggestion
******************
Profile HPU models
******************
```

File: docs/source-pytorch/tuning/profiler_advanced.rst [hunk 14]
```diff
@@ -72,3 +72,115 @@ Make sure the code is running while you are trying to capture the traces. It wil
...
+To understand the cost of each PyTorch operation, use the :class:`~pytorch_lightning.profilers.hpu.HPUProfiler` built on top of the `PyTorch profiler <https://pytorch.org/docs/1.12/profiler.html#torch-profiler>`__.
+
+.. code-block:: python
+
+    from pytorch_lightning.profilers import HPUProfiler
+
+    profiler = HPUProfiler()
+    trainer = Trainer(profiler=profiler)
```
Comment (Borda) [inline docs/source-pytorch/tuning/profiler_advanced.rst:95]:
```suggestion
    trainer = Trainer(profiler=HPUProfiler())
```
Or do I need to have a profiler instance to call summary?

File: src/pytorch_lightning/profilers/hpu.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+            # otherwise, this will raise a `segmentation fault`.
+            if self._should_override_schedule():
+                warning_cache.warn(
+                    "The HPUProfiler default schedule will be overridden as there is not enough "
+                    "steps to properly record traces."
+                )
+                self._schedule = None
+                self.profiler.schedule = torch.profiler.profiler._default_schedule_fn
+
+            def on_trace_ready(profiler: _PROFILER) -> None:
```
Comment (Borda) [inline src/pytorch_lightning/profilers/hpu.py:125]: Does this need to be inside another method? or can be defined as static

File: tests/tests_pytorch/profilers/test_hpu_profiler.py [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+from pytorch_lightning.profilers import AdvancedProfiler, HPUProfiler, SimpleProfiler
+from tests_pytorch.helpers.runif import RunIf
+
+
+class TestHPUProfiler:
+    def setup_method(self):
+        shutil.rmtree("profiler_logs", ignore_errors=True)
+
+    def teardown_method(self):
+        shutil.rmtree("profiler_logs", ignore_errors=True)
```
Comment (Borda) [inline tests/tests_pytorch/profilers/test_hpu_profiler.py:34]: Can we rather use plain function as everywhere else and, instead of this cleaning use temp folders?

File: src/pytorch_lightning/profilers/hpu.py (hunk 15 above)
Comment (kaushikb11) [reply src/pytorch_lightning/profilers/hpu.py:125]: can't be static, need access to self

File: docs/source-pytorch/tuning/profiler_advanced.rst [hunk 17]
```diff
@@ -72,3 +72,114 @@ Make sure the code is running while you are trying to capture the traces. It wil
...
+To understand the cost of each PyTorch operation, use the :class:`~pytorch_lightning.profilers.hpu.HPUProfiler` built on top of the `PyTorch profiler <https://pytorch.org/docs/1.12/profiler.html#torch-profiler>`__.
+
+.. code-block:: python
+
+    from pytorch_lightning.profilers import HPUProfiler
+
+    trainer = Trainer(profiler=HPUProfiler())
+
+The profiler will dump a trace file for each profiler step, ``training_step``, ``backward``, ``validation_step``, ``test_step``, and ``predict_step`` by default.
+The user can provide ``PyTorchProfiler(record_functions={...})`` to extend the scope of profiled functions.
```
Comment (rohitgr7) [inline docs/source-pytorch/tuning/profiler_advanced.rst:97]:
```suggestion
The user can provide ``HPUProfiler(record_functions={...})`` to extend the scope of profiled functions.
```

File: docs/source-pytorch/tuning/profiler_advanced.rst [hunk 18]
```diff
@@ -72,3 +72,114 @@ Make sure the code is running while you are trying to capture the traces. It wil
...
+Find bottlenecks in your code on HPU (advanced)
+################################################
+**Audience**: Users who want to profile their HPU models to find bottlenecks and improve performance.
+
+----
+
+******************
+Profile HPU models
+******************
+To understand the cost of each PyTorch operation, use the :class:`~pytorch_lightning.profilers.hpu.HPUProfiler` built on top of the `PyTorch profiler <https://pytorch.org/docs/1.12/profiler.html#torch-profiler>`__.
```
Comment (rohitgr7) [inline docs/source-pytorch/tuning/profiler_advanced.rst:88]:
```suggestion
To understand the cost of each PyTorch operation, use the :class:`~pytorch_lightning.profilers.hpu.HPUProfiler` built on top of the :class:`~pytorch_lightning.profilers.pytorch.PyTorchProfiler`.
```

File: docs/source-pytorch/tuning/profiler_advanced.rst [hunk 19]
```diff
@@ -72,3 +72,114 @@ Make sure the code is running while you are trying to capture the traces. It wil
...
+    is_complex             0.06%            0.961us          0.06%            0.961us          0.481us
+    stride                 0.03%            0.517us          0.03%            0.517us          0.517us
+    ---------------------  ---------------  ---------------  ---------------  ---------------  ---------------
+    Self CPU time total: 1.681ms
+
+.. note::
+    Since HPUProfiler extends PyTorch Profiler, when using the HPUProfiler, wall clock time will not be representative of the true wall clock time.
+    This is due to forcing profiled operations to be measured synchronously, when many HPU ops happen asynchronously.
+    It is recommended to use this Profiler to find bottlenecks/breakdowns, however for end to end wall clock time use
+    the ``SimpleProfiler``.
```
Comment (rohitgr7) [inline docs/source-pytorch/tuning/profiler_advanced.rst:135]:
```suggestion
    the :class:`~pytorch_lightning.profilers.simple.SimpleProfiler`.
```

File: examples/pl_hpu/hpu_profiler_example.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+        )
+        _accelerator = "hpu"
+    else:
+        _profiler = PyTorchProfiler()
+        _accelerator = "cpu"
+        warnings.warn(
+            f"""_KINETO_AVAILABLE is {_KINETO_AVAILABLE}. Continuing with
+                      profiler="PyTorchProfiler"
+                      accelerator="{_accelerator}" """
+        )
```
Comment (rohitgr7) [inline examples/pl_hpu/hpu_profiler_example.py:92]:
```suggestion
        profiler = HPUProfiler(
            record_shapes=True,
            profile_memory=True,
            with_stack=True,
            with_flops=True,
            with_modules=True,
        )
        accelerator = "hpu"
    else:
        profiler = PyTorchProfiler()
        accelerator = "cpu"
        warnings.warn(
            f"""_KINETO_AVAILABLE is {_KINETO_AVAILABLE}. Continuing with
                      profiler="PyTorchProfiler"
                      accelerator="{accelerator}" """
        )
```

File: examples/pl_hpu/hpu_profiler_example.py [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+
+    trainer = Trainer(
+        profiler=_profiler,
+        accelerator=_accelerator,
```
Comment (rohitgr7) [inline examples/pl_hpu/hpu_profiler_example.py:96]:
```suggestion
        profiler=profiler,
        accelerator=accelerator,
```

File: src/pytorch_lightning/profiler/__init__.py [hunk 22]
```diff
@@ -28,4 +29,5 @@
     "PyTorchProfiler",
     "SimpleProfiler",
     "XLAProfiler",
+    "HPUProfiler",
```
Comment (rohitgr7) [inline src/pytorch_lightning/profiler/__init__.py:32]: let's not add it here. this is deprecated sub-dir

File: src/pytorch_lightning/profilers/__init__.py [hunk 23]
```diff
@@ -25,4 +26,5 @@
     "PyTorchProfiler",
     "SimpleProfiler",
     "XLAProfiler",
+    "HPUProfiler",
```
Comment (rohitgr7) [inline src/pytorch_lightning/profilers/__init__.py:29]: can you put it in sorted order?

File: src/pytorch_lightning/profilers/hpu.py [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+                removes the limit completely.
+
+            sort_by_key: Attribute used to sort entries. By default
+                they are printed in the same order as they were registered.
+                Valid keys include: ``cpu_time``, ``cpu_time_total``,
+                ``cpu_memory_usage``, ``self_cpu_memory_usage``, ``count``.
+
+            record_module_names: Whether to add module names while recording autograd operation.
+
+            profiler_kwargs: Keyword arguments for the PyTorch profiler. This depends on your PyTorch version
```
Comment (rohitgr7) [inline src/pytorch_lightning/profilers/hpu.py:75]:
```suggestion
            \**profiler_kwargs: Keyword arguments for the PyTorch profiler. This depends on your PyTorch version
```

File: tests/tests_pytorch/profilers/test_hpu_profiler.py (hunk 16 above)
Comment (rohitgr7) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:34]: yeah, I agree we should use plain functions here. the logs are in the tmp folder.

File: tests/tests_pytorch/profilers/test_hpu_profiler.py [hunk 25]
```diff
@@ -0,0 +1,224 @@
...
+            return 1
+        assert hpus <= HPUAccelerator.auto_device_count(), "More hpu devices asked than present"
+        assert hpus == 1 or hpus % 8 == 0
+        return hpus
+
+    @RunIf(hpu=True)
+    def test_hpu_simple_profiler_instances(self, tmpdir, get_device_count):
+        model = BoringModel()
+        trainer = Trainer(
+            profiler="simple", accelerator="hpu", devices=get_device_count, max_epochs=1, default_root_dir=tmpdir
```
Comment (rohitgr7) [inline tests/tests_pytorch/profilers/test_hpu_profiler.py:50]: can we use fast_dev_run in all of these tests?

File: tests/tests_pytorch/profilers/test_hpu_profiler.py [hunk 26]
```diff
@@ -0,0 +1,224 @@
...
+        )
+        assert isinstance(trainer.profiler, SimpleProfiler)
+        trainer.fit(model)
```
Comment (rohitgr7) [inline tests/tests_pytorch/profilers/test_hpu_profiler.py:53]: this one doesn't need a fit call

File: tests/tests_pytorch/profilers/test_hpu_profiler.py [hunk 27]
```diff
@@ -0,0 +1,224 @@
...
+        expected = {f"{stage}-profiler.txt" for stage in ("fit", "validate", "test", "predict")}
+        assert actual == expected
+        for file in list(os.listdir(profiler.dirpath)):
+            assert os.path.getsize(os.path.join(profiler.dirpath, file)) > 0
+
+    @RunIf(hpu=True)
+    def test_hpu_advanced_profiler_instances(self, tmpdir, get_device_count):
+        model = BoringModel()
+        trainer = Trainer(
+            profiler="advanced", accelerator="hpu", devices=get_device_count, max_epochs=1, default_root_dir=tmpdir
```
Comment (rohitgr7) [inline tests/tests_pytorch/profilers/test_hpu_profiler.py:79]: doesn't need a fit call

File: tests/tests_pytorch/profilers/test_hpu_profiler.py [hunk 28]
```diff
@@ -0,0 +1,224 @@
...
+        for file in list(os.listdir(profiler.dirpath)):
+            assert os.path.getsize(os.path.join(profiler.dirpath, file)) > 0
+
+    @RunIf(hpu=True)
+    def test_hpu_pytorch_profiler_instances(tmpdir):
+        model = BoringModel()
+
+        trainer = Trainer(profiler="hpu", accelerator="hpu", devices=1, max_epochs=1)
+        assert isinstance(trainer.profiler, HPUProfiler)
+        trainer.fit(model)
```
Comment (rohitgr7) [inline tests/tests_pytorch/profilers/test_hpu_profiler.py:110]: doesn't need a fit call

File: src/pytorch_lightning/profiler/__init__.py (hunk 22 above)
Comment (jerome-habana) [reply src/pytorch_lightning/profiler/__init__.py:32]: I got mypy / flake errors

File: tests/tests_pytorch/profilers/test_hpu_profiler.py (hunk 25 above)
Comment (jerome-habana) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:50]: I dont see a limitation

File: tests/tests_pytorch/profilers/test_hpu_profiler.py (hunk 26 above)
Comment (jerome-habana) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:53]: doesn't hurt to check if fit works with profiling

File: tests/tests_pytorch/profilers/test_hpu_profiler.py (hunk 25 above)
Comment (rohitgr7) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:50]: but it increases the ci time.. which we can avoid.

File: tests/tests_pytorch/profilers/test_hpu_profiler.py (hunk 16 above)
Comment (kaushikb11) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:34]: @jerome-habana yup, let's remove it.

File: tests/tests_pytorch/profilers/test_hpu_profiler.py [hunk 29]
```diff
@@ -0,0 +1,235 @@
...
+                raise Exception("Could not find event cpu_op in trace")
+            for event_duration in event_duration_arr:
+                assert event_duration >= 0
+
+    @RunIf(hpu=True)
+    def test_hpu_trace_event_hpu_op(tmpdir):
+        # Run model and prep json
+        model = BoringModel()
+
+        trainer = Trainer(accelerator="hpu", devices=1, max_epochs=1, profiler=HPUProfiler())
```
Comment (rohitgr7) [inline tests/tests_pytorch/profilers/test_hpu_profiler.py:158]: fast_dev_run

File: tests/tests_pytorch/profilers/test_hpu_profiler.py [hunk 30]
```diff
@@ -0,0 +1,235 @@
...
+                raise Exception("Could not find event hpu_op in trace")
+            for event_duration in event_duration_arr:
+                assert event_duration >= 0
+
+    @RunIf(hpu=True)
+    def test_hpu_trace_event_hpu_meta_op(tmpdir):
+        # Run model and prep json
+        model = BoringModel()
+
+        trainer = Trainer(accelerator="hpu", devices=1, max_epochs=1, profiler=HPUProfiler())
```
Comment (rohitgr7) [inline tests/tests_pytorch/profilers/test_hpu_profiler.py:186]: fast_dev_run

File: tests/tests_pytorch/profilers/test_hpu_profiler.py [hunk 31]
```diff
@@ -0,0 +1,235 @@
...
+            if len(event_duration_arr) == 0:
+                raise Exception("Could not find event hpu_meta_op in trace")
+            for event_duration in event_duration_arr:
+                assert event_duration >= 0
+
+    @RunIf(hpu=True)
+    def test_hpu_trace_event_kernel(tmpdir):
+        # Run model and prep json
+        model = BoringModel()
+        trainer = Trainer(accelerator="hpu", devices=1, max_epochs=1, profiler=HPUProfiler())
```
Comment (rohitgr7) [inline tests/tests_pytorch/profilers/test_hpu_profiler.py:214]: fast_dev_run

File: tests/tests_pytorch/profilers/test_hpu_profiler.py (hunk 16 above)
Comment (jerome-habana) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:34]: Removing this will have issues with multi-process runs. There is no single unified directory to hold the logs

File: tests/tests_pytorch/profilers/test_hpu_profiler.py [hunk 32]
```diff
@@ -0,0 +1,235 @@
...
+
+        trainer = Trainer(profiler="hpu", accelerator="hpu", devices=1, max_epochs=1, fast_dev_run=True)
+        assert isinstance(trainer.profiler, HPUProfiler)
+
+    @RunIf(hpu=True)
+    def test_hpu_trace_event_cpu_op(tmpdir):
+        # Run model and prep json
+        model = BoringModel()
+
+        trainer = Trainer(accelerator="hpu", devices=1, max_epochs=1, profiler=HPUProfiler())
```
[the remark this answers is not in the document]
Comment (jerome-habana) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:123]: fast_dev in these places, doesn't generate the log files. So we wont be able to parse. https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#fast-dev-run . It disables all events/ logging

File: tests/tests_pytorch/profilers/test_hpu_profiler.py (hunk 29 above)
Comment (jerome-habana) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:152]: ditto

File: tests/tests_pytorch/profilers/test_hpu_profiler.py (hunk 30 above)
Comment (jerome-habana) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:180]: ditto

File: tests/tests_pytorch/profilers/test_hpu_profiler.py (hunk 31 above)
Comment (jerome-habana) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:208]: ditto

File: tests/tests_pytorch/profilers/test_hpu_profiler.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (rohitgr7) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:123]: you can limit the batches
```
limit_train_batches=1
limit_val_batches=1
max_epochs=1
```

[the remark this answers is not in the document]
Comment (jerome-habana) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:123]: Will this limiting enable logging too ?

[the remark this answers is not in the document]
Comment (rohitgr7) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:123]: yes, this just limits the batches per epoch

[the remark this answers is not in the document]
Comment (rohitgr7) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:123]: which document? did you remove `fast_dev_run` after adding `limit_*_batches`?

[the remark this answers is not in the document]
Comment (jerome-habana) [reply tests/tests_pytorch/profilers/test_hpu_profiler.py:123]: @rohitgr7 https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#fast-dev-run


'Disables all loggers.

Disables passing logged metrics to loggers.'

We can work it around by adding some more changes in the test by changing log parsing, settign val to none etc. My proposal would be to leave the test as the focus is to showcase hpu profier

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/profilers/hpu.py [hunk 33]
```diff
@@ -0,0 +1,151 @@
...
+
+if _KINETO_AVAILABLE:
+    from torch.profiler import profile, ProfilerActivity, tensorboard_trace_handler
+
+log = logging.getLogger(__name__)
+
+_PROFILER = profile
+
+
+class HPUProfiler(PyTorchProfiler):
```
Comment (tchaton) [inline src/pytorch_lightning/profilers/hpu.py:36]: Hey @jerome-habana, great work there!

I wonder if we couldn't do slightly better there.

It seems the main reason for overriding was the error message. Maybe we could re-organize the PyTorch Profiler slightly to accommodate accelerator-specific activity better.

Comment (jerome-habana) [reply src/pytorch_lightning/profilers/hpu.py:36]: Agree @tchaton . This would need some change to PT profiler with hpu support. We are working to enable some event mechanics there too.

**Comment by Borda** [comment]
as discussed with @lantiga this shall be implemented via integrations :rabbit:
